src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingAssignmentOperator.cs

```diff
@@ -14,35 +16,143 @@ public override BoundNode VisitNullCoalescingAssignmentOperator(BoundNullCoalesc
+            return result;
+
+            BoundExpression rewriteNullCoalscingAssignmentStandard()
+            {
+                // Now that LHS is transformed with temporaries, we rewrite this node into a coalesce expression:
+                // lhsRead ?? (transformedLHS = loweredRight)
+
+                // transformedLHS = loweredRight
+                // isCompoundAssignment is only used for dynamic scenarios, and we want those scenarios to treat this like a standard assignment.
+                // See CodeGenNullCoalescingAssignmentTests.CoalescingAssignment_DynamicRuntimeCastFailure, which will fail if
+                // isCompoundAssignment is set to true. It will fail to throw a runtime binder cast exception.
+                BoundExpression assignment = MakeAssignmentOperator(syntax, transformedLHS, loweredRight, node.LeftOperand.Type, used: true, isChecked: false, isCompoundAssignment: false);
 
-            // lhsRead ?? (transformedLHS = loweredRight)
-            BoundExpression conditionalExpression = MakeNullCoalescingOperator(syntax, lhsRead, assignment, Conversion.Identity, BoundNullCoalescingOperatorResultKind.LeftType, node.LeftOperand.Type);
+                // lhsRead ?? (transformedLHS = loweredRight)
+                BoundExpression conditionalExpression = MakeNullCoalescingOperator(syntax, lhsRead, assignment, Conversion.Identity, BoundNullCoalescingOperatorResultKind.LeftType, node.LeftOperand.Type);
 
-            BoundExpression result = (temps.Count == 0 && stores.Count == 0) ?
+                return (temps.Count == 0 && stores.Count == 0) ?
                     conditionalExpression :
                     new BoundSequence(
                         syntax,
-                        temps.ToImmutable(),
-                        stores.ToImmutable(),
+                        temps.ToImmutableAndFree(),
+                        stores.ToImmutableAndFree(),
                         conditionalExpression,
                         conditionalExpression.Type);
+            }
 
-            temps.Free();
-            stores.Free();
+            // Rewrites the null coalescing operator in the case where the result type is the underlying
+            // non-nullable value type of the left side
+            BoundExpression rewriteNullCoalescingAssignmentForValueType()
+            {
+                Debug.Assert(node.LeftOperand.Type.IsNullableType());
+                Debug.Assert(node.Type.Equals(node.RightOperand.Type));
 
-            return result;
+                // We lower the expression to this form:
+                //
+                // var tmp = lhsRead.GetValueOrDefault();
+                // if (!lhsRead.HasValue) { tmp = loweredRight; transformedAssignment = tmp; }
+                // tmp
+                //
+                // Except that a is only evaluated once.
+
+                var leftOperand = node.LeftOperand;
+                if (!TryGetNullableMethod(leftOperand.Syntax,
```

Comment:
>if (!TryGetNullableMethod(leftOperand.Syntax, [](start = 16, length = 45)

Do we have tests for those missing methods?

Reply:
See `CodeGenNullCoalescingAssignmentTests.MissingMembers`

---
In reply to: [293090659](https://github.com/dotnet/roslyn/pull/36329#discussion_r293090659) [](ancestors = 293090659)